### PR TITLE
tengine: init init at v2.2.2

### PR DESCRIPTION
**Pull request by Izorkin**
###### Motivation for this change
Add web server tengine.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/servers/http/tengine/default.nix [hunk 1]
```diff
@@ -0,0 +1,82 @@
+{ stdenv, fetchurl, openssl, zlib, pcre, libxml2, libxslt
+, gd, geoip
+, withDebug ? false
+, withMail ? false
+, withIPv6 ? false
```
Comment (xeji) [inline pkgs/servers/http/tengine/default.nix]: Why disable IPv6? I believe that any server application should support IPv6 by default.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -19954,6 +19954,11 @@ with pkgs;
 
   teeworlds = callPackage ../games/teeworlds { };
 
+  tengine = callPackage ../servers/http/tengine {
+    stdenv = overrideCC stdenv gcc6;
+    modules = [ nginxModules.rtmp nginxModules.dav nginxModules.moreheaders nginxModules.modsecurity-nginx];
```
Comment (xeji) [inline pkgs/top-level/all-packages.nix]: `with nginxModules; [ rtmp dav ... ]`

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -19954,6 +19954,11 @@ with pkgs;
 
   teeworlds = callPackage ../games/teeworlds { };
 
+  tengine = callPackage ../servers/http/tengine {
+    stdenv = overrideCC stdenv gcc6;
```
Comment (xeji) [inline pkgs/top-level/all-packages.nix]: Is this necessary? Did you try to build with default gcc7 ?

File: pkgs/servers/http/tengine/default.nix (hunk 1 above)
Comment (Izorkin) [reply pkgs/servers/http/tengine/default.nix]: Ok, fixed.

File: pkgs/top-level/all-packages.nix (hunk 3 above)
Comment (Izorkin) [reply pkgs/top-level/all-packages.nix]: Error build with gcc7.

**Comment by xeji** [comment]
@Izorkin i added a commit to fix the build with gcc7 so we do not need to use gcc6. Please test.

**Comment by Izorkin** [comment]
Thanks. Works.

**Comment by xeji** [comment]
@GrahamcOfBorg build tengine

**Comment by xeji** [comment]
aarch64 failure can be fixed later by someone familiar with the platform.

**Pull request merged**